[Search] Add Azure Machine Learning Skill

Mohit-Chakraborty opened the pull request.

Mohit-Chakraborty committed:
Update generated code to add `AmlSkill`

Mohit-Chakraborty committed:
Update public APi

Add AmlSkill to roundtripping test

Mohit-Chakraborty committed:
Merge remote-tracking branch 'origin/main' into mohitc/AmlSkill

Mohit-Chakraborty committed:
Merge with origin/main

Mohit-Chakraborty committed:
Revert "Merge with origin/main"

This reverts commit f1c2bd6e40b8a6196ad335da7eb903616081c968.

azure-sdk commented:
API changes have been detected in `Azure.Search.Documents`. You can review API changes [here](https://apiview.dev/Assemblies/Review/2f5726d5a3ca4e71a67a379ef9b94e8f?diffOnly=True&diffRevisionId=9a36f44415474201b28fedd9c6a5901a)

**API changes**
```diff
+     public class AmlSkill : SearchIndexerSkill {
+         public AmlSkill(IEnumerable<InputFieldMappingEntry> inputs, IEnumerable<OutputFieldMappingEntry> outputs);
+         public int? DegreeOfParallelism { get; set; }
+         public string Key { get; set; }
+         public string Region { get; set; }
+         public string ResourceId { get; set; }
+         public TimeSpan? Timeout { get; set; }
+         public string Uri { get; set; }
+     }
```

Mohit-Chakraborty committed:
Update session record for RoundtripAllSkills test

Mohit-Chakraborty committed:
Add transforms to rename properties
Expose appropriate constructors

Mohit-Chakraborty commented:
cc: @tg-msft

azure-sdk commented:
API changes have been detected in `Azure.Search.Documents`. You can review API changes [here](https://apiview.dev/Assemblies/Review/2f5726d5a3ca4e71a67a379ef9b94e8f?diffOnly=True&diffRevisionId=9a36f44415474201b28fedd9c6a5901a)

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -325,6 +325,18 @@ public partial class SimpleFieldAttribute : System.Attribute
+    {
+        public AmlSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, string resourceId) { }
+        public AmlSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri) { }
+        public AmlSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri, string authenticationKey) { }
```
Should we have params for all these?  Or just the setters below?

Mohit-Chakraborty replied:
These are the required combinations to create a valid skill creation request.
I did it as per the requirements mentioned in the [skill document](https://docs.microsoft.com/en-us/azure/search/cognitive-search-aml-skill#what-skill-parameters-to-use).

![image](https://user-images.githubusercontent.com/8271806/151079845-812cb505-eed1-4e3b-9c0d-0496ae314572.png)

tg-msft replied:
Then do all the required things need to be settable?

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -325,6 +325,18 @@ public partial class SimpleFieldAttribute : System.Attribute
+        public AmlSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, string resourceId) { }
+        public AmlSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri) { }
+        public AmlSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri, string authenticationKey) { }
+        public string AuthenticationKey { get { throw null; } set { } }
```
Should this be exposed?  Or should we treat this like a credential that can be rolled?  We had the same conversation per https://apiview.dev/Assemblies/Review/f291bdd4488149748454e2e2fa18a821#Azure.Search.Documents.Indexes.Models.KnowledgeStore and I'd either copy that if it GA'ed or we can talk about fixing both to be more like credentials.

Mohit-Chakraborty replied:
`KnowledgeStore` [has GAed](https://github.com/Azure/azure-sdk-for-net/blob/0ebd32c0359bcd7a20b1708683c323a0b26b618a/sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs#L890) and it has getter and setter on its `StorageConnectionString`.

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -325,6 +325,18 @@ public partial class SimpleFieldAttribute : System.Attribute
+        public string AuthenticationKey { get { throw null; } set { } }
+        public int? DegreeOfParallelism { get { throw null; } set { } }
+        public string Region { get { throw null; } set { } }
+        public string ResourceId { get { throw null; } set { } }
```
Is this an https://apiview.dev/Assemblies/Review/77bdbcef49a3465fa96fac3fb745bf57#Azure.Core.ResourceIdentifier?

tg-msft replied:
+@KrzysztofCwalina and @m-nash in case they want to weigh in on the use of these new shared types for ResourceId/Region.

m-nash replied:
The [response in the recording](https://github.com/Azure/azure-sdk-for-net/pull/26597/files#diff-1710cfe8188cf5793ac17a39d6db68365df758d3c73f9cf8be93430d2058fcabR452) has null for the ID, is this dummy data?  If this is an ARM resource id then we can definitely use the new class.

Mohit-Chakraborty replied:
It is an ARM resource id.

KrzysztofCwalina replied:
If we can use them here, we should :-)

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -325,6 +325,18 @@ public partial class SimpleFieldAttribute : System.Attribute
+        public AmlSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri, string authenticationKey) { }
+        public string AuthenticationKey { get { throw null; } set { } }
+        public int? DegreeOfParallelism { get { throw null; } set { } }
+        public string Region { get { throw null; } set { } }
```
Is this an https://apiview.dev/Assemblies/Review/77bdbcef49a3465fa96fac3fb745bf57#Azure.Core.AzureLocation?

tg-msft replied:
Actually, do we want to rename the property to `Location` (in just .NET) so it's `public AzureLocation? Location { get; }` rather than `public AzureLocation? Region { get; }`?

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -325,6 +325,18 @@ public partial class SimpleFieldAttribute : System.Attribute
+        public string Region { get { throw null; } set { } }
+        public string ResourceId { get { throw null; } set { } }
+        public System.Uri ScoringUri { get { throw null; } set { } }
+        public System.TimeSpan? Timeout { get { throw null; } set { } }
```
What's this the timeout for?  Prediction?  Training?

Mohit-Chakraborty replied:
This is the 'timeout for the http client making the API call'.

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -325,6 +325,18 @@ public partial class SimpleFieldAttribute : System.Attribute
 }
 namespace Azure.Search.Documents.Indexes.Models
 {
+    public partial class AmlSkill : Azure.Search.Documents.Indexes.Models.SearchIndexerSkill
```
`Aml` isn't a great prefix to most of the world.  Could we call this something liker `MachineLearningSkill` or maybe `AzureMachineLearningSkill`?

tg-msft reviewed (approved):
Looks good overall, but we should chat about naming across languages.

Mohit-Chakraborty commented:
> Looks good overall, but we should chat about naming across languages.

Agree. I see it as a 2-step process. It is easier for me to understand and present the (proposed) changes in SDK code. Coupled with the fact that merging a PR in the REST spec repo is a longer process, this approach seems more efficient. With an approval here, we can update the REST spec and consume in all languages and remove the customizations here.

Mohit-Chakraborty committed:
Take model rename changes from REST spec
Make required properties as get-only
Change type of Location -> Azure.Core.AzureLocation
Change type of ResourceId -> Azure.Core.ResourceIdentifier

azure-sdk commented:
API changes have been detected in `Azure.Search.Documents`. You can review API changes [here](https://apiview.dev/Assemblies/Review/2f5726d5a3ca4e71a67a379ef9b94e8f?diffOnly=True&diffRevisionId=9a36f44415474201b28fedd9c6a5901a)

Mohit-Chakraborty committed:
Mark `AzureMachineLearningSkill.Location` as nullable

azure-sdk commented:
API changes have been detected in `Azure.Search.Documents`. You can review API changes [here](https://apiview.dev/Assemblies/Review/2f5726d5a3ca4e71a67a379ef9b94e8f?diffOnly=True&diffRevisionId=9a36f44415474201b28fedd9c6a5901a)

Mohit-Chakraborty commented:
Can you please take another look - @tg-msft?

Mohit-Chakraborty committed:
Update changelog

tg-msft commented on sdk/search/Azure.Search.Documents/CHANGELOG.md:
```diff
@@ -1,14 +1,10 @@
-### Bugs Fixed
-
-### Other Changes
+- Added `Unk` as an `OcrSkillLanguage` value. The values are used to set the default language code for the [OCR cognitive skill](https://docs.microsoft.com/azure/search/cognitive-search-skill-ocr).
```
Is there a public API called `Unk`?  I'm not seeing it in this PR, but if it exists should it be called `Unknown` instead?

Mohit-Chakraborty replied:
Yes, this was [added recently](https://github.com/Azure/azure-sdk-for-net/pull/25812) to the SDK. It was missing from the REST spec as a supported value.

tg-msft replied:
We should chat about this one during standup today because I've got a few questions: partly how this relates to a `null` value for the extensible enum, partly the discoverability of `Unk` vs `Unknown` _(since I definitely didn't realize that when I approved the original pr :smile:)_, and partly a concern that https://iso639-3.sil.org/code/unk is already a language code _(though it's ISO 639-3 and I think all the other values here are just ISO 639-1)_.

Mohit-Chakraborty replied:
Sounds good.
[The documentation](https://docs.microsoft.com/en-us/azure/search/cognitive-search-skill-ocr#skill-parameters) says -
![image](https://user-images.githubusercontent.com/8271806/152065573-8b9debb7-128e-4c55-a949-bc445366c418.png)

We can switch to saying `Unknown` to make the meaning clear, we have to remember to do it for every instance going forward. Currently, other skills do not use 'unknown' as a don't-care-figure-it-out option.

tg-msft commented on sdk/search/Azure.Search.Documents/CHANGELOG.md:
```diff
@@ -1,14 +1,10 @@
-
-### Other Changes
+- Added `Unk` as an `OcrSkillLanguage` value. The values are used to set the default language code for the [OCR cognitive skill](https://docs.microsoft.com/azure/search/cognitive-search-skill-ocr).
+- Support for [`AzureMachineLearningSkill`](https://docs.microsoft.com/azure/search/cognitive-search-aml-skill). The AML skill allows you to extend AI enrichment with a custom [Azure Machine Learning](https://docs.microsoft.com/azure/machine-learning/overview-what-is-azure-machine-learning) (AML) model. Once an AML model is [trained and deployed](https://docs.microsoft.com/azure/machine-learning/ ...
```
This is an awesome explanation.  Can we use it as a doc comment too so it shows up in our ref docs?

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -351,6 +351,18 @@ public partial class AsciiFoldingTokenFilter : Azure.Search.Documents.Indexes.Mo
     }
+    public partial class AzureMachineLearningSkill : Azure.Search.Documents.Indexes.Models.SearchIndexerSkill
+    {
+        public AzureMachineLearningSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, string resourceId) { }
+        public AzureMachineLearningSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri) { }
+        public AzureMachineLearningSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri, string authenticationKey) { }
```
Per https://docs.microsoft.com/en-us/azure/search/cognitive-search-aml-skill#what-skill-parameters-to-use, should we instead make these look more like:
```suggestion
        public AzureMachineLearningSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, ResourceIdentifier resourceId, AzureLocation? location = default) { }
        public AzureMachineLearningSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri, string authenticationKey = default) { }
```
since `resourceId` shouldn't be a string, `Location` only matters when you're using `ResourceId`, and optional params would make it easier to pick between the different overloads?

Mohit-Chakraborty replied:
This is much better. Thanks for the suggestion.

tg-msft commented on sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:
```diff
@@ -351,6 +351,18 @@ public partial class AsciiFoldingTokenFilter : Azure.Search.Documents.Indexes.Mo
+        public AzureMachineLearningSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, System.Uri scoringUri, string authenticationKey) { }
+        public string AuthenticationKey { get { throw null; } }
+        public int? DegreeOfParallelism { get { throw null; } set { } }
+        public Azure.Core.AzureLocation? Location { get { throw null; } set { } }
```
If we move this to the .ctor, can this become readonly?

Mohit-Chakraborty replied:
Yes.

tg-msft commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/AzureMachineLearningSkill.cs:
```diff
@@ -0,0 +1,319 @@
+    /// </summary>
+    public partial class AzureMachineLearningSkill : IUtf8JsonSerializable
+    {
+        /// <summary> (Required for key authentication) The key for the AML service. </summary>
```
I don't think I've seen our ref docs start with parenthesized text, it feels like it'd be better to say what it is before we jump into how it's required, and it might be helpful to explain what it's used with?
```suggestion
        /// <summary>The key for the Azure Machine Learning service.  This is required for key-based authentication along with <ref-magic>ScoringUri</ref-magic>.</summary>
```

tg-msft commented on sdk/search/Azure.Search.Documents/src/Indexes/Models/AzureMachineLearningSkill.cs:
```diff
@@ -0,0 +1,319 @@
+            ODataType = "#Microsoft.Skills.Custom.AmlSkill";
+        }
+
+        void global::Azure.Core.IUtf8JsonSerializable.Write(Utf8JsonWriter writer)
```
Why do we need to do all this?

Mohit-Chakraborty replied:
CodeGen did not generate compilable code for `ResourceIdentifier` and `AzureLocation`.
I have opened [an issue](https://github.com/Azure/autorest.csharp/issues/1933).

tg-msft replied:
Oh, I see.  In that case, can we do the "use an internal property as a makeshift converter" trick so we don't have to own updating the serialization code whenever the swagger changes?  Imagine something like:

```C#
        [CodeGenMember("Region")]
        internal string RawLocation
        {
            get => Location?.ToString();
            set => Location = (value == null ? null : new AzureLocation(value));
        }
        
        public AzureLocation? Location { get; private set; }
```

tg-msft reviewed (approved):
I think there are a few small things that could be tweaked, but it looks great overall.

Mohit-Chakraborty committed:
Refactor ctors on AmlSkill
Add hyperlinks on comments

azure-sdk commented:
API changes have been detected in `Azure.Search.Documents`. You can review API changes [here](https://apiview.dev/Assemblies/Review/2f5726d5a3ca4e71a67a379ef9b94e8f?diffOnly=True&diffRevisionId=9a36f44415474201b28fedd9c6a5901a)

**API changes**
```diff
+     public class AzureMachineLearningSkill : SearchIndexerSkill {
+         public AzureMachineLearningSkill(IEnumerable<InputFieldMappingEntry> inputs, IEnumerable<OutputFieldMappingEntry> outputs, Uri scoringUri, string authenticationKey = null);
+         public AzureMachineLearningSkill(IEnumerable<InputFieldMappingEntry> inputs, IEnumerable<OutputFieldMappingEntry> outputs, ResourceIdentifier resourceId, AzureLocation? location = null);
+         public string AuthenticationKey { get; }
+         public int? DegreeOfParallelism { get; set; }
+         public AzureLocation? Location { get; }
+         public ResourceIdentifier ResourceId { get; }
+         public Uri ScoringUri { get; }
+         public TimeSpan? Timeout { get; set; }
+     }
```

Mohit-Chakraborty committed:
Redefine properties to support (de)serialization
Define internal-only variations of 2 properties so the generated (de)serialization code works without being manually overridden.
